// @contrast/protobuf-api/protobuf/service_api_grpc_pb.js
// GENERATED CODE -- DO NOT EDIT!

'use strict';
var grpc = require('grpc');
var dtm_pb = require('./dtm_pb.js');
var settings_pb = require('./settings_pb.js');

function serialize_contrast_api_dtm_Message(arg) {
  if (!(arg instanceof dtm_pb.Message)) {
    throw new Error('Expected argument of type contrast.api.dtm.Message');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_contrast_api_dtm_Message(buffer_arg) {
  return dtm_pb.Message.deserializeBinary(new Uint8Array(buffer_arg));
}

function serialize_contrast_api_settings_AgentSettings(arg) {
  if (!(arg instanceof settings_pb.AgentSettings)) {
    throw new Error('Expected argument of type contrast.api.settings.AgentSettings');
  }
  return Buffer.from(arg.serializeBinary());
}

function deserialize_contrast_api_settings_AgentSettings(buffer_arg) {
  return settings_pb.AgentSettings.deserializeBinary(new Uint8Array(buffer_arg));
}


var ContrastService = exports.ContrastService = {
  sendAgentMessage: {
    path: '/contrast.apis.Contrast/SendAgentMessage',
    requestStream: false,
    responseStream: false,
    requestType: dtm_pb.Message,
    responseType: settings_pb.AgentSettings,
    requestSerialize: serialize_contrast_api_dtm_Message,
    requestDeserialize: deserialize_contrast_api_dtm_Message,
    responseSerialize: serialize_contrast_api_settings_AgentSettings,
    responseDeserialize: deserialize_contrast_api_settings_AgentSettings,
  },
};

exports.ContrastClient = grpc.makeGenericClientConstructor(ContrastService);
